File: mobile/civic_engagement_app/lib/src/features/citizen/domain/entities/notification.dart
import 'package:equatable/equatable.dart';

/// Notification types as per backend API Section 12.3
enum NotificationType {
  policyActivated('POLICY_ACTIVATED'),
  policyAutoActivated('POLICY_AUTO_ACTIVATED'),
  policyClosed('POLICY_CLOSED'),
  policyAutoClosed('POLICY_AUTO_CLOSED'),
  policyExtended('POLICY_EXTENDED'),
  policyLifecycle('POLICY_LIFECYCLE'),
  associateAssigned('ASSOCIATE_ASSIGNED'),
  associatePermissionsUpdated('ASSOCIATE_PERMISSIONS_UPDATED'),
  associateRevoked('ASSOCIATE_REVOKED'),
  messageReceived('MESSAGE_RECEIVED'),
  commentReply('COMMENT_REPLY'),
  commentFlagged('COMMENT_FLAGGED'),
  commentAppeal('COMMENT_APPEAL'),
  appealResolved('APPEAL_RESOLVED'),
  voteSurge('VOTE_SURGE'),
  ratingDrop('RATING_DROP'),
  emergingTopic('EMERGING_TOPIC'),
  plannerApproved('PLANNER_APPROVED'),
  plannerRequestApproved('PLANNER_REQUEST_APPROVED'),
  accountDeletionConfirmed('ACCOUNT_DELETION_CONFIRMED'),
  unknown('UNKNOWN');

  final String value;
  const NotificationType(this.value);

  static NotificationType fromString(String value) {
    return NotificationType.values.firstWhere(
      (type) => type.value == value,
      orElse: () => NotificationType.unknown,
    );
  }
}

/// Notification severity levels
enum NotificationSeverity {
  info('info'),
  warning('warning'),
  critical('critical');

  final String value;
  const NotificationSeverity(this.value);

  static NotificationSeverity fromString(String value) {
    return NotificationSeverity.values.firstWhere(
      (severity) => severity.value == value,
      orElse: () => NotificationSeverity.info,
    );
  }
}

/// Notification source
enum NotificationSource {
  system('system'),
  alert('alert');

  final String value;
  const NotificationSource(this.value);

  static NotificationSource fromString(String value) {
    return NotificationSource.values.firstWhere(
      (source) => source.value == value,
      orElse: () => NotificationSource.system,
    );
  }
}

/// Notification entity as per backend API Section 12.4
class NotificationEntity extends Equatable {
  final String id;
  final String userId;
  final String userRole;
  final NotificationType type;
  final String title;
  final String message;
  final Map<String, dynamic>? data;
  final bool read;
  final NotificationSeverity severity;
  final NotificationSource source;
  final DateTime createdAt;

  const NotificationEntity({
    required this.id,
    required this.userId,
    required this.userRole,
    required this.type,
    required this.title,
    required this.message,
    this.data,
    required this.read,
    required this.severity,
    required this.source,
    required this.createdAt,
  });

  /// Check if this is a smart alert
  bool get isSmartAlert => source == NotificationSource.alert;

  /// Check if this is a critical notification
  bool get isCritical => severity == NotificationSeverity.critical;

  /// Check if this is a warning
  bool get isWarning => severity == NotificationSeverity.warning;

  /// Get policy ID from data if available
  String? get policyId => data?['policyId'] as String?;

  /// Get comment ID from data if available
  String? get commentId => data?['commentId'] as String?;

  @override
  List<Object?> get props => [
        id,
        userId,
        userRole,
        type,
        title,
        message,
        data,
        read,
        severity,
        source,
        createdAt,
      ];
}

File: backend/src/models/Notification.js
const mongoose = require("mongoose");

const notificationSchema = new mongoose.Schema(
  {
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
      required: true,
    },
    userRole: {
      type: String,
      enum: ["citizen", "planner", "admin"],
      required: true,
    },
    type: {
      type: String,
      enum: [
        // Policy events
        "POLICY_ACTIVATED",
        "POLICY_CLOSED",
        "POLICY_EXTENDED",

        // Delegation / Associates
        "ASSOCIATE_INVITED",
        "ASSOCIATE_ACCEPTED",
        "ASSOCIATE_REJECTED",
        "ASSOCIATE_PERMISSIONS_UPDATED",
        "ASSOCIATE_REVOKED",
        "ASSOCIATE_SELF_REVOKED",
        "ASSOCIATE_INVITATION_CANCELLED",
        "INVITATION_EXPIRED", // owner
        "INVITATION_EXPIRING_SOON", // associate (receiver)
        "INVITATION_EXPIRED_RECEIVER", // associate (receiver)

        // Messaging
        "MESSAGE_RECEIVED",

        // Comments
        "COMMENT_REPLY",
        "COMMENT_FLAGGED",
        "COMMENT_APPEAL",
        "APPEAL_RESOLVED",

        // Smart alerts
        "VOTE_SURGE",
        "RATING_DROP",
        "EMERGING_TOPIC",
        "PLANNER_APPROVED",
        "PLANNER_REQUEST_CREATED",
        "PLANNER_APPEAL_APPROVED",
      ],
      required: true,
    },
    title: { type: String, required: true },
    message: { type: String, required: true },
    data: { type: Object, default: {} },
    read: { type: Boolean, default: false },
    severity: {
      type: String,
      enum: ["info", "warning", "critical"],
      default: "info",
    },
    source: { type: String, enum: ["system", "alert"], default: "system" },
  },
  {
    timestamps: true, // replaces createdAt/updatedAt
  },
);

// Indexes
notificationSchema.index({ userId: 1, createdAt: -1 });
notificationSchema.index({ read: 1 });
notificationSchema.index({ userId: 1, type: 1 });
notificationSchema.index({ createdAt: 1 }); // for cleanup jobs

module.exports = mongoose.model("Notification", notificationSchema);
